daemon/daemon_utils.go

```diff
@@ -35,20 +41,50 @@ func initialRuntime(baseDir string, runtimes map[string]types.Runtime) error {
 
 	// create script for runtime who has args
 	for name, r := range runtimes {
-		if len(r.RuntimeArgs) == 0 {
-			continue
-		}
-
-		script := filepath.Join(dir, name)
 		if r.Path == "" {
 			r.Path = name
 		}
-		data := fmt.Sprintf("#!/bin/sh\n%s %s $@\n", r.Path, strings.Join(r.RuntimeArgs, " "))
 
-		if err := ioutil.WriteFile(script, []byte(data), runtimeScriptPerm); err != nil {
-			return fmt.Errorf("failed to create runtime script %s: %s", script, err)
+		// setup a fake path
+		if len(r.RuntimeArgs) != 0 {
+			data := fmt.Sprintf("#!/bin/sh\n%s %s $@\n", r.Path, strings.Join(r.RuntimeArgs, " "))
+			r.Path = filepath.Join(dir, name)
+
+			if err := ioutil.WriteFile(r.Path, []byte(data), runtimeScriptPerm); err != nil {
+				return fmt.Errorf("failed to create runtime script %s: %s", r.Path, err)
+			}
+		}
+
+		if r.Type == "" {
+			r.Type = ctrd.RuntimeTypeRuncV1
+		}
+
+		// convert general json map to specific options type
+		b, err := json.Marshal(r.Options)
+		if err != nil {
+			return fmt.Errorf("failed to marshal options, runtime: %s: %v", name, err)
+		}
+		options := getRuntimeOptionsType(r.Type)
+		if err := json.Unmarshal(b, options); err != nil {
+			return fmt.Errorf("failed to unmarshal to type %+v: %v", options, err)
 		}
+
+		r.Options = options
```

Comment:
I think this is not correct, runtime type is different from runtime which we add in config. For runtime type, it can be pass by cri , we do not need to add them into config, and validate it.
As we get more runtime concept, we can remove the runtime validate, just add parameter here (it can be disscussed)

Reply:
What passed from CRI is also an alias name of actual runtime type and runtime options.

ContainerManager side should maintain a map (name -> runtime config), and CRI would use the name to choose the actual containerd runtime config.

Reply:
This can be decided by containerd, why use a alias name ?